File: src/old_communication_layer/lib/translation.py
import requests
import sys

def translate(text, context=None, port=11434) -> str:
    '''
    The following function makes a request to the model API
    The function accepts a string of Chinese text
    as well as a string of context in either English or Chinese
    The context is used to improve the translation
    context is optional and defaults to None
    The function returns a string of English text
    '''
    # The URL for the model API
    url = f"http://127.0.0.1:{port}/api/generate/"

    # Check if the text is valid
    if not is_valid(text):
        # Log the error
        print("Invalid text: " + text)
        # Exit the program
        sys.exit(1)

    if context is not None:
        # The data to be sent to the model API
        data = {
                "model": "llama2",
                "prompt": text,
                "context": context
                }

    else:
        # The data to be sent to the model API
        data = {
                "model": "llama2",
                "prompt": text
                }

    # Make a request to the model API
    response = requests.post(url, data=data)

    # Return the English translation
    return response.text

def is_valid(text) -> bool:
    '''
    The following function accepts a string of Chinese text
    It ensures that the text is valid UTF-8
    The function returns a boolean
    '''
    try:
        text.encode("utf-8")
        return True
    except:
        return False

File: src/gollama/utils/translation.go
package utils

// Handles the translation of text
// Makes http requests to the ollama api
// Returns a string of translated text

// Author: Korben Tompkin

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// Struct to hold the response from the ollama api
type Response struct {
	Response string `json:"response"`
}

// Translate a string of text
func Translate(text string, sourceLanguage string, targetLanguage string,
	customPrompt string, customModel string, opts Options) (string, error) {
	// Array of response strings
	var responses []string

	// System prompt to be sent to the ollama api
	var prompt string

	// Format the system prompt with the source and target languages if no custom prompt is provided
	if customPrompt == "" {
		prompt = fmt.Sprintf("Please translate the given %s sentence into formal and academic %s without any comment or discussion. Do not include any additional discussion or comment.", sourceLanguage, targetLanguage)
	} else {
		prompt = customPrompt
	}

	// Model to be sent to the ollama api, default to llama2 if no custom model is provided
	var model string
	if customModel == "" {
		model = "llama2"
	} else {
		model = customModel
	}

	// Create the request url
	url := "http://127.0.0.1:11434/api/generate/"

	// Format the request body
	data, err := json.Marshal(map[string]string{
		"model":  model,
		"prompt": text,
		"system": prompt,
	})

	// Marshal the options object
	optsData, err := json.Marshal(map[string]interface{}{
		"temperature": opts.Temperature,
		"num_ctx":     opts.Context,
	})

	// Add the options to the request body inside the options object
	data = []byte(strings.Replace(string(data), "}", ",\"options\":"+string(optsData)+"}", 1))

	// Create the request
	req, err := http.NewRequest("POST", url, strings.NewReader(string(data)))
	if err != nil {
		return "", err
	}

	// Set the request headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	// Send the request
	client := &http.Client{}

	// Read the first response so we can get the status code
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}

	reader := bufio.NewReader(resp.Body)
	for {
		line, err := reader.ReadBytes('\n')
		if err != nil {
			if err == io.EOF {
				break
			}
			return "", err
		}

		// unmarshal the response
		var response Response
		err = json.Unmarshal(line, &response)
		if err != nil {
			return "", err
		}

		// Add the response to the array
		responses = append(responses, response.Response)
	}

	// turn the response into a string
	result := strings.Join(responses, "")

	// Return the translated text
	return result, nil
}
